### PR TITLE
ORC-1709: Upgrade GitHub Action `setup-java` to v4 and use built-in cache feature

### PR DESCRIPTION
**Pull request by sullis**
### What changes were proposed in this pull request?

upgrade GitHub Action setup-java from v3 to v4

Also:  setup-java supports a "cache" attribute.  This allows us to remove the "cache" action from the workflow.

The "cache" attribute is documented in the setup-java README:
https://github.com/actions/setup-java/blob/main/README.md

image

### Why are the changes needed?

v4 is the latest version of setup-java

### How was this patch tested?

GitHub Actions workflow.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -67,18 +67,12 @@ jobs:
...
-        key: ${{ matrix.java }}-maven-${{ hashFiles('**/pom.xml') }}
-        restore-keys: |
-          ${{ matrix.java }}-maven-
     - name: Install Java ${{ matrix.java }}
-      uses: actions/setup-java@v3
+      uses: actions/setup-java@v4
       with:
         distribution: zulu
         java-version: ${{ matrix.java }}
+        cache: 'maven'
```
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:75]: Could you elaborate more about this new feature in the PR description?

Comment (sullis) [reply .github/workflows/build_and_test.yml:75]: Ack. I have updated the PR description.

**Comment by dongjoon-hyun** [comment]
cc @williamhyun

**Review by williamhyun — COMMENTED** [review]
```
Failed to save: Unable to reserve cache with key setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf48815b63ee075e7d3fc5a9c1f392f, another job may be creating this cache. More details: Cache already exists. Scope: refs/pull/1925/merge, Key: setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf48815b63ee075e7d3fc5a9c1f392f, Version: 3d88543681947283edf8997cb89a71c1522b2e39bc75c36ae780bd628eb8559d
Cache saved with the key: setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf[4](https://github.com/apache/orc/actions/runs/8990923635/job/24698981632#step:9:4)8815b63ee075e7d3fc5a9c1f392f
```
@sullis
Is this working correctly?

**Comment by dongjoon-hyun** [comment]
Gentle ping for the above comment, @sullis .

If we are not sure, let's revert `cache: 'maven'` part and recover the old one. I believe the version bump is enough.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @sullis .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Since we need to verify the validity, let's wait for the author's confirmation.

**Comment by sullis** [comment]
> ```
> Failed to save: Unable to reserve cache with key setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf48815b63ee075e7d3fc5a9c1f392f, another job may be creating this cache. More details: Cache already exists. Scope: refs/pull/1925/merge, Key: setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf48815b63ee075e7d3fc5a9c1f392f, Version: 3d88543681947283edf8997cb89a71c1522b2e39bc75c36ae780bd628eb8559d
> Cache saved with the key: setup-java-Linux-maven-fc278625f3aefc0088ce0c0131f0800ddbf[4](https://github.com/apache/orc/actions/runs/8990923635/job/24698981632#step:9:4)8815b63ee075e7d3fc5a9c1f392f
> ```
> 
> @sullis Is this working correctly?


Interesting.   I'd like to know if "Failed to save: Unable to reserve cache with key " is repeatable.

Can you re-run the workflows?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Sorry for a long delay. I got a chance to revisit this PR today.

Merged to main.

**Comment by dongjoon-hyun** [comment]
Do you have a JIRA ID? I can assign ORC-1709 to you.

**Pull request closed without merging**